swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt

```diff
@@ -0,0 +1,165 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.swabbie.notifications
+
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.kork.lock.LockManager
+import com.netflix.spinnaker.swabbie.LockingService
+import com.netflix.spinnaker.swabbie.discovery.DiscoveryAware
+import com.netflix.spinnaker.swabbie.events.OwnerNotifiedEvent
+import com.netflix.spinnaker.swabbie.events.UnMarkResourceEvent
+import com.netflix.spinnaker.swabbie.model.MarkedResource
+import com.netflix.spinnaker.swabbie.model.NotificationInfo
+import com.netflix.spinnaker.swabbie.model.WorkConfiguration
+import com.netflix.spinnaker.swabbie.repository.ResourceTrackingRepository
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.stereotype.Component
+import java.time.Clock
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+
+/**
+ * A scheduled notification sender
+ */
+@Component
+@ConditionalOnExpression("\${swabbie.enabled:true}")
+class NotificationSender(
+  private val lockingService: LockingService,
+  private val resourceTrackingRepository: ResourceTrackingRepository,
+  private val notifier: Notifier,
+  private val clock: Clock,
+  private val applicationEventPublisher: ApplicationEventPublisher,
+  private val notificationQueue: NotificationQueue,
+  private val registry: Registry,
+  private val workConfigurations: List<WorkConfiguration>
+) : DiscoveryAware() {
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+  private val notificationsId = registry.createId("swabbie.notifications")
+
+  private val lockOptions = LockManager.LockOptions()
+    .withLockName(lockingService.ownerName)
+    .withMaximumLockDuration(lockingService.swabbieMaxLockDuration)
+
+  /**
+   * Periodically sends notifications to resolved resource owners
+   * The frequency of this scheduled action is cron based and configurable
+   * Defaults to every day at 9am.
+   */
+  @Scheduled(cron = "\${swabbie.notification.cron.schedule:0 0 9 1/1 * ?}")
+  fun sendNotifications() {
+    if (!isUp()) {
+      return
+    }
+
+    lockingService.acquireLock(lockOptions) {
+      val resources = resourceTrackingRepository.getMarkedResources()
+        .filter {
+          it.notificationInfo == null
+        }
+
+      groupRelatedResourcesThenNotify(resources)
+    }
+  }
+
+  /**
+   * @return Marked resources to notify on
+   */
```

Comment:
I'd like a method level comment here explaining why we need to group resources and then how we're grouping and notifying. I find this method a little hard to follow.